test/runner.js

```diff
@@ -10,87 +9,138 @@ require( 'babel-register' );
+			if ( /test\/?$/.test( filePath ) ) {
+				matches = matches.concat( glob.sync( '*.@(js|jsx)', globOptions ) );
+			}
+
+			return memo.concat( memo, matches );
+		}
+
+		return memo.concat( memo, filePath );
+	}, [] );
+
+	if ( program.nodeTotal > 1 ) {
+		testFiles = testFiles.filter( ( file, index ) => {
+			return index % program.nodeTotal === program.nodeIndex;
+		} );
+	}
 
-if ( process.env.CIRCLECI ) {
-	debug( 'Hello Circle!' );
-	// give circle more time by default because containers are slow
-	// why 10 seconds? a guess.
-	mocha.suite.timeout( 10000 );
-}
+	return testFiles;
+};
+
+const getMocha = function() {
+	/* These internal dependencies are not at the top of the file
+	 * because we need to re-require them every time we invalidate the cache.
+	 *
+	 * We invalidate the cache because Mocha has trouble running a test-suite
+	 * more than once:  https://github.com/mochajs/mocha/issues/995
+	 */
+	Object.keys( require.cache ).forEach( ( key ) => {
+		delete require.cache[ key ];
+	} );
+	const boot = require( './boot-test' );
 
-mocha.suite.beforeAll( boot.before );
-mocha.suite.afterAll( boot.after );
+	const mocha = new Mocha( {
+		ui: 'bdd',
+		reporter: program.reporter
+	} );
 
-files = program.args.length ? program.args : [ process.env.TEST_ROOT ];
-files = files.reduce( ( memo, filePath ) => {
-	// Validate test root matches specified file paths
-	if ( ! filePath.startsWith( process.env.TEST_ROOT ) ) {
-		console.warn(
-			chalk.red.bold( 'WARNING:' ),
-			chalk.yellow( 'Invalid argument passed to test runner. Paths must match test root `' + process.env.TEST_ROOT + '`.' )
-		);
-		console.warn( ' - ' + filePath + '\n' );
+	mocha.suite.beforeAll( boot.before );
+	mocha.suite.afterAll( boot.after );
 
-		return memo;
+	if ( program.grep ) {
+		mocha.grep( new RegExp( program.grep ) );
 	}
 
-	// Append individual file argument
-	if ( /\.jsx?$/i.test( filePath ) ) {
-		return memo.concat( filePath );
+	if ( process.env.CIRCLECI ) {
+		debug( 'Hello Circle!' );
+		// give circle more time by default because containers are slow
+		// why 10 seconds? a guess.
+		mocha.suite.timeout( 10000 );
 	}
 
-	// Determine whether argument already includes intended test directory,
-	// or if we should recursively search for test directories.
-	let globPattern = '*.@(js|jsx)';
-	if ( ! /\/test\/?$/.test( filePath ) ) {
-		globPattern = path.join( '**/test', globPattern );
-	}
+	mocha.files = getTestFiles();
+	return mocha;
+};
+
+let runner = null;
+const runMocha = () => {
+	console.log( chalk.green( 'Running tests at: ' ),
+		chalk.yellow( new Date() ) );
 
-	// Append discovered files from glob result
-	return memo.concat( glob.sync( path.join( filePath, globPattern ) ) );
-}, [] );
+	runner = getMocha().run( ( failures ) => {
+		runner = null;
 
-if ( program.nodeTotal > 1 ) {
-	files = files.filter( ( file, index ) => {
-		return index % program.nodeTotal === program.nodeIndex;
+		// Inform CI of failures
+		process.on( 'exit', () => {
```

Comment:
We should move it to the end of the script. Otherwise we add another even listener each time we rerun tests.